rasterio/_warp.pyx

```diff
@@ -223,6 +223,7 @@ def _reproject(
         num_threads=1,
         warp_mem_limit=0,
         working_data_type=0,
+        src_geoloc_array=None,
```

Comment:
A few usability notes (context in GDAL [RFC #4](https://gdal.org/development/rfc/rfc4_geolocate.html)):

* `kwargs` should disallow the use of `METHOD=GEOLOC_ARRAY` and `SRC_GEOLOC_ARRAY`.
* Is the intention to support `PIXEL_STEP` and `LINE_STEP` via `kwargs`? These seem like key parameters for some users. `PIXEL_OFFSET` and `LINE_OFFSET` seem less useful, but worth supporting for completeness.
* It is possible to assign a geolocation array to the target image using the `DST_GEOLOC_ARRAY` transformer option. If `kwargs` is passed along correctly it seems like this should work if the caller configures things correctly, but it may be worth thinking about whether or not this should be more officially supported, or explicitly disallowed.
* It is possible for the input image to have `GEOLOCATION` metadata set. This is a special metadata domain like `RPC`, and normally points to an external dataset. Another useful pattern is to construct a geolocation array, attach `GEOLOCATION` metadata, and omit the `*_DATASET` metadata items. This dataset can then be applied to any dataset of the right shape without having to manage geolocation metadata on the primary image dataset. Is Rasterio's stance that this metadata will be ignored, and users must manually load it and craft an appropriate call to `reproject()`?
* Rasterio must be able to read and write `GEOLOCATION` metadata, otherwise the library cannot be used to craft an image whose geolocation is provided by a geolocation array. I think I confirmed this works, but IMO should be intentionally tested.
* A geolocation can include an optional `Z_BAND`, although I suspect this is ignored by the geolocation array transformer, and exists to provide a mechanism to include elevation for use elsewhere.